rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,205 @@
+* All query string, path, fragements are contained in a fully formed `url` property
+
+### Origins
+
+Where present, the `origin` rest parameter in this service for will be either
+an IPv4 address, IPv6 address, or an identifier.  This parameter will specify
+the source location of the ultimately originating request.  This means the
+origin of the request to the Queue from the worker and not the request to the
+object service.  If the origin is an IPv4 or IPv6, a mapping of the address and
+IP will occur to find the backing storage.  If the origin is not an IPv4 or
+IPv6 address and is identical to a set of known and mappable identifiers (e.g.
+`s3_us-west-2`), then that will be used to map the token to find the backing
+storage.
+
+### Creation
+
+```
+PUT /artifacts/:name
+PATCH /artifacts/:name
+```
+
+The `PUT` endpoint will take an origin and name.  It will return a `200` with a
+document that contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the artifact in the shape:
+
+```javascript
+{
+  origin: '10.10.10.10' || 's3_us-west-2',
+  contentType: 'application/json',
+  contentLength: 1234,
+  contentSha256: 'abcdef1234',
+  transferLength: 1234,
+  transferSha256: 'abcdef1234',
+  contentEncdoing: 
+  expiration: new Date(),
+}
+```
+
+This endpoint will return a list of HTTP requests which must all be run.
+
+The `PATCH` endpoint will be sent without a request body and will be used to
+perform any post-upload actions.  These include things like the commit step on
+an S3 multipart upload.  This endpoint would be run by the Queue and not by the
+uploader, and the Queue would also need a complimentary method for workers to
+call when they complete the upload.  This is what is currently done with the
+`blob` storage type in the queue.
+
+### Deletetion
+
+```
+DELETE /artifacts/:name
+DELETE /caches/artifacts/:name
+```
+
+This service has at least one copy of each stored file.  Any copies above 
+
+Delete all stored copies of the named object if using first endpoint.  The
+second endpoint would attempt delete all of the cached copies in the various
+locations.  The cached copies are those copies which are created in other
+regions or services to provide more local copies of each object.  This cache
+purging must return a `202 Accepted` response and not `200/204` due to the
+nature of how caches work in the real world.
+
+### Retreival
+
+```
+GET /artifacts/:name[?max_time=30]
+```
+
+This endpoint returns a 302 redirect to the location of the object which is
+closest to the request's IP address.  If needed, this will initiate a copy from
+the original backing store into a different storage location.  It will wait up
+to a configurable amount of time before redirecting to a non-optimal copy.
```

Comment:
There is an `origin` involved here, too, right?  Can that be provided as a query parameter?